pkg/config/config.go

```diff
@@ -0,0 +1,125 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	serverconfig "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/brancz/kube-rbac-proxy/pkg/authn"
+	"github.com/brancz/kube-rbac-proxy/pkg/authz"
+	"github.com/brancz/kube-rbac-proxy/pkg/proxy"
+)
+
+type KubeRBACProxyConfig struct {
+	SecureServing *serverconfig.SecureServingInfo
+
+	KubeRBACProxyInfo *KubeRBACProxyInfo
+}
+
+type KubeRBACProxyInfo struct {
+	InsecureListenAddress string // DEPRECATED
+	HealthzPath           string
+
+	UpstreamURL      *url.URL
+	UpstreamForceH2C bool
+
+	UpstreamTransport http.RoundTripper
+
+	Auth *proxy.Config
+
+	KubeClient *kubernetes.Clientset
+
+	AllowPaths  []string
+	IgnorePaths []string
+}
+
+func NewConfig() *KubeRBACProxyConfig {
+	return &KubeRBACProxyConfig{
+		SecureServing: &serverconfig.SecureServingInfo{},
+		KubeRBACProxyInfo: &KubeRBACProxyInfo{
+			Auth: &proxy.Config{
+				Authentication: &authn.AuthnConfig{
+					X509:   &authn.X509Config{},
+					Header: &authn.AuthnHeaderConfig{},
+					OIDC:   &authn.OIDCConfig{},
+					Token:  &authn.TokenConfig{},
+				},
+				Authorization: &authz.Config{},
+			},
+		},
+	}
+}
+
+func (i *KubeRBACProxyInfo) SetUpstreamTransport(upstreamCAPool *x509.CertPool, upstreamClientCertPath, upstreamClientKeyPath string) error {
+	if upstreamCAPool == nil {
+		i.UpstreamTransport = http.DefaultTransport
+		return nil
+	}
+
+	var certKeyPair tls.Certificate
+	if len(upstreamClientCertPath) > 0 {
+		var err error
+		certKeyPair, err = tls.LoadX509KeyPair(upstreamClientCertPath, upstreamClientKeyPath)
+		if err != nil {
+			return fmt.Errorf("failed to read upstream client cert/key: %w", err)
+		}
+	}
+
+	// http.Transport sourced from go 1.10.7
+	transport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			RootCAs: upstreamCAPool,
+		},
+	}
+
+	if certKeyPair.Certificate != nil {
+		transport.TLSClientConfig.Certificates = []tls.Certificate{certKeyPair}
+	}
+
+	i.UpstreamTransport = transport
+	return nil
+}
+
+func (i *KubeRBACProxyConfig) GetClientCAProvider() (dynamiccertificates.CAContentProvider, error) {
```

Comment:
Yeah, at this point I am curious if it wouldn't be better to name it something like `DownstreamClientCA`. In reference to the explicit `UpstreamClientCA`.

Reply:
maybe we should rename the `UpstreamClientCA` to just `UpstreamCA` instead

Reply:
We could do so as well. But Client itself is not specific enough as we receive client connections and we create client connections.